### PR TITLE
Add Weighted Gradient-Enhanced MAE Loss Function

### PR DESCRIPTION
**Pull request by mkeutgen**
**Summary**
Implements decomposed_mae_gradient_weighted() - a configurable loss function that combines standard MAE with spatial gradient penalties to preserve mesoscale ocean features (eddies, fronts) during training.

**Problem**
Pure MAE loss produces smooth predictions that lose sharp oceanographic features. Models achieve good R² scores but generate unrealistic, overly-diffused fields that lack physical structure

**Solution**
Add tunable gradient penalty term:
`Loss = MAE(pred, target) + α × gradient_penalty(pred, target)
`
Where α (gradient_weight) controls the accuracy vs. sharpness tradeoff:

α = 0.1: Conservative balance (recommended starting point)
α = 0.25: More aggressive feature preservation
α = 0.5: Better sharpness

**Implementation Details**

Per-channel loss computation for multi-variable training
Spatial gradients computed via finite differences (∂/∂x, ∂/∂y)
Gradient penalties averaged and weighted before combination

**Benefits**

Preserves mesoscale eddy coherence during autoregressive rollouts
Maintains fronts and boundaries
Tunable via a single hyperparameter
Computationally efficient (simple finite differences)

### CODE REVIEW

File: src/ocean_emulators/aggregator/inference/reduced.py [hunk 1]
```diff
@@ -140,9 +140,14 @@ def record(self, target: torch.Tensor, gen: torch.Tensor, i_time_start: int):
                 "target and gen must have the same shape, got "
                 f"{target.shape} and {gen.shape}"
             )
+        # Note(mkeutgen):  this prevents a dimension mismatch error
+        target = target.squeeze()
```
Comment (alxmrs) [inline src/ocean_emulators/aggregator/inference/reduced.py:144]: Would you mind removing this commit from this branch/PR? It seems to be left over from a change unrelated to the primary purpose of this PR. 

The easiest way may be to make a new branch from main and then to cherry pick the latest commit in this branch. 

Another option could be to do an interactive rebase (there are some instructions in our contributing guide) and to drop the first two commits. Then, when the change and commit history looks good, I recommend doing a force push with lease (see the contrib guide).

**Review by alxmrs — COMMENTED** [review]
A quick note! A full review will come soon as I find time.

File: src/ocean_emulators/utils/loss.py [hunk 2]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
 
+def decomposed_mae_gradient_weighted(
+    pred: torch.Tensor, 
+    target: torch.Tensor, 
+    wet: torch.Tensor,
+    gradient_weight: float = 0.1
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:79]: Let's omit adding a default kwarg, (except in test code). In my experience, it helps prevent subtle configuration bugs.

File: src/ocean_emulators/utils/loss.py [hunk 3]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
...
+) -> torch.Tensor:
+    """
+    MAE loss with WEIGHTED spatial gradient matching penalty.
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:82]: This docstring is excellent. Thanks for explaining what this loss method does. 

Is there a paper associated with this method? It would be nice to add a citation (even just a URL).

File: src/ocean_emulators/utils/loss.py [hunk 4]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
     combined = (mse + mae) / 2
     return combined.mean(dim=(0, 2, 3))
 
+def decomposed_mae_gradient_weighted(
+    pred: torch.Tensor, 
+    target: torch.Tensor, 
+    wet: torch.Tensor,
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:78]: Let's use the jaxtyping value here called Grid from our constants file.

File: src/ocean_emulators/utils/loss.py [hunk 5]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
...
+    
+    Recommended starting values:
+    - α = 0.05: Very conservative, prioritize accuracy
+    - α = 0.1:  Conservative, good balance 
+    - α = 0.25: Moderate, more sharpness 
+    - α = 0.5:  Aggressive sharpening
+    - α = 1.0:  Equal weighting 
+    
+    Args:
+        pred: Predicted tensor [batch, channels, height, width]
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:99]: Optional nit: let's end each line of the docstring with a full stop ".".

File: src/ocean_emulators/utils/loss.py [hunk 6]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
...
+        target: Target tensor [batch, channels, height, width]
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:100]: Optional: consider incorporating this into the type via jaxtyping.

File: src/ocean_emulators/utils/loss.py [hunk 7]
```diff
@@ -72,6 +72,68 @@ def decomposed_mse_mae(
...
+        Loss per channel [channels]
+    """
+    pred = pred * wet
+    target = target * wet
+    
+    # MAE term (main accuracy objective)
+    mae_loss = F.l1_loss(pred, target, reduction="none")
+    mae_per_channel = mae_loss.mean(dim=(0, 2, 3))
+    
+    # Gradient penalty: Match spatial gradients
```
Comment (alxmrs) [inline src/ocean_emulators/utils/loss.py:114]: Can you help me understand, why do we do this differencing across height and width?

Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py:114]: To me, this reminds me of some kind of simple edge detector, see

https://cseweb.ucsd.edu/classes/sp16/cse152-a/lec7.pdf

**Review by alxmrs — COMMENTED** [review]
More notes and a question.

File: src/ocean_emulators/utils/loss.py (hunk 7 above)
Comment (alxmrs) [reply src/ocean_emulators/utils/loss.py:114]: Sorry, I read your README just now. This makes more sense!

Comment (mkeutgen) [reply src/ocean_emulators/utils/loss.py:114]: Hi Alex thanks for your very useful feedback, I've created a new clean PR I hope this new PR is better :) I still need to find a paper about gradient loss, I've not found one yet
